Refactor samd21

keestux opened the pull request:
This is the fourth attempt to reorganize cpu/samd21 so that it will be possible to add new boards like SODAQ Autonomo and Arduino Zero.

There are 11 commits which do the following steps

* move the current samd21 include files to sam21_common (only CMSIS)
* update the samr21 CMSIS from ASF 3.21
* add the samd21 CMSIS from ASF 3.30
* move i2c to sam21_common
* refactor samd21 spi and uart

Notice that there is no new "cpu" and a start is made to move common code to sam21_common. The only thing I want to mention is that samd21 boards now need  a few extras in their Makefile.include, like

    export CPU = samd21
    export SAM0_FAM = samd21
    export CPU_MODEL = samd21j18a   # Not sure if this is used
    CFLAGS += -D__SAMD21J18A__  # or whatever variant you have

This is already taken care of in samr21-xpro.

@PeterKietzmann was kind enough to already test these changes on a samr21-xpro and confirmed that everything continues to work.

PeterKietzmann commented:
Yes I did test UART, I2C and SPI with this branch on a samr21-xpro and everything worked as expected. Also note that Murdocks output looks pretty nice. It just complains about whitespace errors in CMSIS headers. How do we handle that?

kaspar030 commented:
> Murdocks output looks pretty nice. It just complains about whitespace errors in CMSIS headers. How do we handle that?

Add the dirs to the ```EXCLUDE_PATTERNS``` in ```doc/doxygen/riot.doxyfile```.

kaspar030 commented on cpu/sam21_common/Makefile.include:
```diff
@@ -1,6 +1,8 @@
+CFLAGS += -DCPU_FAM_$(shell echo $(CPU_FAM) | tr 'a-z-' 'A-Z_')
+
+# Define the SAM0 family so we can differentiate between them in the code
+CFLAGS += -DSAM0_FAM_$(shell echo $(SAM0_FAM) | tr 'a-z-' 'A-Z_')
```
Any reason to not use ```CPU_FAM``` instead of ```SAM0_FAM```? Seems redundant.

keestux replied:
I looked at it, but the problem is that `CPU_FAM` is currently set in `cpu/samd21/Makefile.include`. And that one is shared for samd21 and samr21 boards. We need a parameter to distinguish between these two.

Notice that for samr21-xpro `SAM0_FAM` is set to `samr21`.

That why I came up with `SAM0_FAM` which must be set in the board configuration.

kaspar030 replied:
> I looked at it, but the problem is that `CPU_FAM` is currently set in `cpu/samd21/Makefile.include`. And that one is shared for samd21 and samr21 boards. We need a parameter to distinguish between these two.

Why? We don't have to use the dispatching from CMSIS.

We currently support one samd21 CPU (SAMR21G18A), and that number might
rise to 5-10 in the mid-term.

IMHO we should just select the right cmsis header directly (in cpu_conf.h).

Reduces "-DCPU_FAM=samd21 -DSAM0_FAM=samr21 -D__SAMR21G18A__
-DCPU_FAM_SAMD21 -DSAM0_FAM_SAMR21" by 40%. ;)

keestux replied:
Why do you pick on me? I'd say, look at your own contribution of cpu/saml21. It has the whole set of cmsis for saml21.

> IMHO we should just select the right cmsis header directly (in cpu_conf.h).

You can't. Not without adding a define like, eh, `SAM0_FAM_xyz`. If you had let me move samd21 to samr21 and add the 'real' samd21, then the three `samd21`, `samr21`, `saml21` would all have had the same structure. And `SAM0_FAM` wouldn't be needed.

gebart commented:
@keestux Thank you for your patience with this tedious task.

Please fix the whitespace errors (trailing whitespaces) in the CMSIS headers as a separate commit.
`find cpu/samr21/include -name '*.h' -exec sed 's/\s*$//' +` should fix it I think.

kaspar030 commented:
Are the samd21 CMSIS headers used anywhere with this PR? If not please cut them out of this PR for better reviewability, and add them together with the first board using them (autonomo?)

kaspar030 commented:
> @keestux Thank you for your patience with this tedious task.

+1!

keestux commented:
@kaspar030 About the samd21 CMSIS files. I rather keep them in.
I needed this sequence of commits to see if samr21-xpro and sodaq-autonomo could coexist.

kaspar030 commented:
> Why do you pick on me? I'd say, look at your own contribution of cpu/saml21. It has the whole set of cmsis for saml21.

I'm sorry if you have the impression that I'm picking on you. That's definitely not my intend. Feel invited for a drink the moment we meet in person! :)

>> IMHO we should just select the right cmsis header directly (in cpu_conf.h).
> You can't. Not without adding a define like, eh, SAM0_FAM_xyz. If you had let me move samd21 to samr21 and add the 'real' samd21, then the three samd21, samr21, saml21 would all have had the same structure. And SAM0_FAM wouldn't be needed.

Well, I think we shouldn't let the vendor headers influence too much on how we best organize our platforms.

The samr21 is just a variant of samd21, even if Atmel is shipping a whole new include tree. A diff shows that they are practically identical.

Please take a look at https://github.com/keestux/RIOT/pull/1.

In there I've just copied the four samr21 specific files into the samd21 cmsis folder, and added the extra case to samd21.h. Works fine as is, saves 20k lines of duplicated header files and removes the need for SAM0_FAM.

gebart commented:
Den 13 aug. 2016 09:32 skrev "Kaspar Schleiser" <notifications@github.com>:
>>
>> Why do you pick on me? I'd say, look at your own contribution of
cpu/saml21. It has the whole set of cmsis for saml21.
>
> I'm sorry if you have the impression that I'm picking on you. That's
definitely not my intend. Feel invited for a drink the moment we meet in
person! :)
>>>
>>> IMHO we should just select the right cmsis header directly (in
cpu_conf.h).
>>> You can't. Not without adding a define like, eh, SAM0_FAM_xyz. If you
had let me move samd21 to samr21 and add the 'real' samd21, then the three
samd21, samr21, saml21 would all have had the same structure. And SAM0_FAM
wouldn't be needed.
>
> Well, I think we shouldn't let the vendor headers influence too much on
how we best organize our platforms.
>
> The samr21 is just a variant of samd21, even if Atmel is shipping a whole
new include tree. A diff shows that they are practically identical.
>
> Please take a look at keestux#1.
>
> In there I've just copied the four samr21 specific files into the samd21
cmsis folder, and added the extra case to samd21.h. Works fine as is, saves
20k lines of duplicated header files and removes the need for SAM0_FAM.

Keep in mind the maintainability of the files. I don't see any problem with
vendor headers duplicating their content when it reduces our own
maintenence effort, so I don't agree with your opinion, Kaspar. But I just
want this issue to be resolved so that we can focus on more productive
tasks.

>
> —
> You are receiving this because you commented.
>
> Reply to this email directly, view it on GitHub, or mute the thread.

kaspar030 commented:
well, I promise to script the merging of the two cmsis trees the next time one of them needs an update.

-----Original Message-----
From: "Joakim Nohlgård" <notifications@github.com>
To: RIOT-OS/RIOT <RIOT@noreply.github.com>
Cc: Kaspar Schleiser <kaspar@schleiser.de>, Mention <mention@noreply.github.com>
Sent: Sa., 13 Aug. 2016 10:12
Subject: Re: [RIOT-OS/RIOT] Refactor samd21 (#5743)

Den 13 aug. 2016 09:32 skrev "Kaspar Schleiser" <notifications@github.com>:
>>
>> Why do you pick on me? I'd say, look at your own contribution of
cpu/saml21. It has the whole set of cmsis for saml21.
>
> I'm sorry if you have the impression that I'm picking on you. That's
definitely not my intend. Feel invited for a drink the moment we meet in
person! :)
>>>
>>> IMHO we should just select the right cmsis header directly (in
cpu_conf.h).
>>> You can't. Not without adding a define like, eh, SAM0_FAM_xyz. If you
had let me move samd21 to samr21 and add the 'real' samd21, then the three
samd21, samr21, saml21 would all have had the same structure. And SAM0_FAM
wouldn't be needed.
>
> Well, I think we shouldn't let the vendor headers influence too much on
how we best organize our platforms.
>
> The samr21 is just a variant of samd21, even if Atmel is shipping a whole
new include tree. A diff shows that they are practically identical.
>
> Please take a look at keestux#1.
>
> In there I've just copied the four samr21 specific files into the samd21
cmsis folder, and added the extra case to samd21.h. Works fine as is, saves
20k lines of duplicated header files and removes the need for SAM0_FAM.

Keep in mind the maintainability of the files. I don't see any problem with
vendor headers duplicating their content when it reduces our own
maintenence effort, so I don't agree with your opinion, Kaspar. But I just
want this issue to be resolved so that we can focus on more productive
tasks.

>
> —
> You are receiving this because you commented.
>
> Reply to this email directly, view it on GitHub, or mute the thread.


-- 
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub:
https://github.com/RIOT-OS/RIOT/pull/5743#issuecomment-239609206

gebart commented:
Okay, agreed. Let's get this in so that @keestux can get back to working on
the port for the new boards.
@kaspar030 Will you do the review ?

Den 13 aug. 2016 11:20 AM skrev "Kaspar Schleiser" <notifications@github.com
>:

> well, I promise to script the merging of the two cmsis trees the next time
> one of them needs an update.
>
> -----Original Message-----
> From: "Joakim Nohlgård" <notifications@github.com>
> To: RIOT-OS/RIOT <RIOT@noreply.github.com>
> Cc: Kaspar Schleiser <kaspar@schleiser.de>, Mention <
> mention@noreply.github.com>
> Sent: Sa., 13 Aug. 2016 10:12
> Subject: Re: [RIOT-OS/RIOT] Refactor samd21 (#5743)
>
> Den 13 aug. 2016 09:32 skrev "Kaspar Schleiser" <notifications@github.com
> >:
> >>
> >> Why do you pick on me? I'd say, look at your own contribution of
> cpu/saml21. It has the whole set of cmsis for saml21.
> >
> > I'm sorry if you have the impression that I'm picking on you. That's
> definitely not my intend. Feel invited for a drink the moment we meet in
> person! :)
> >>>
> >>> IMHO we should just select the right cmsis header directly (in
> cpu_conf.h).
> >>> You can't. Not without adding a define like, eh, SAM0_FAM_xyz. If you
> had let me move samd21 to samr21 and add the 'real' samd21, then the three
> samd21, samr21, saml21 would all have had the same structure. And SAM0_FAM
> wouldn't be needed.
> >
> > Well, I think we shouldn't let the vendor headers influence too much on
> how we best organize our platforms.
> >
> > The samr21 is just a variant of samd21, even if Atmel is shipping a whole
> new include tree. A diff shows that they are practically identical.
> >
> > Please take a look at keestux#1.
> >
> > In there I've just copied the four samr21 specific files into the samd21
> cmsis folder, and added the extra case to samd21.h. Works fine as is, saves
> 20k lines of duplicated header files and removes the need for SAM0_FAM.
>
> Keep in mind the maintainability of the files. I don't see any problem with
> vendor headers duplicating their content when it reduces our own
> maintenence effort, so I don't agree with your opinion, Kaspar. But I just
> want this issue to be resolved so that we can focus on more productive
> tasks.
>
> >
> > —
> > You are receiving this because you commented.
> >
> > Reply to this email directly, view it on GitHub, or mute the thread.
>
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/RIOT-OS/RIOT/pull/5743#issuecomment-239609206
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/5743#issuecomment-239611665>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AATYQo4P4UpuPqOpbhdXd2KGW0LlPiI9ks5qfYxegaJpZM4Jhh2G>
> .
>

kaspar030 commented:
well, picking on keestux was me already reviewing. :)

as the samr21 is probably our most used board, I'll do more thorough testing. won't get to it until tomorrow, though.

-----Original Message-----
From: "Joakim Nohlgård" <notifications@github.com>
To: RIOT-OS/RIOT <RIOT@noreply.github.com>
Cc: Kaspar Schleiser <kaspar@schleiser.de>, Mention <mention@noreply.github.com>
Sent: Sa., 13 Aug. 2016 11:35
Subject: Re: [RIOT-OS/RIOT] Refactor samd21 (#5743)

Okay, agreed. Let's get this in so that @keestux can get back to working on
the port for the new boards.
@kaspar030 Will you do the review ?

Den 13 aug. 2016 11:20 AM skrev "Kaspar Schleiser" <notifications@github.com
>:

> well, I promise to script the merging of the two cmsis trees the next time
> one of them needs an update.
>
> -----Original Message-----
> From: "Joakim Nohlgård" <notifications@github.com>
> To: RIOT-OS/RIOT <RIOT@noreply.github.com>
> Cc: Kaspar Schleiser <kaspar@schleiser.de>, Mention <
> mention@noreply.github.com>
> Sent: Sa., 13 Aug. 2016 10:12
> Subject: Re: [RIOT-OS/RIOT] Refactor samd21 (#5743)
>
> Den 13 aug. 2016 09:32 skrev "Kaspar Schleiser" <notifications@github.com
> >:
> >>
> >> Why do you pick on me? I'd say, look at your own contribution of
> cpu/saml21. It has the whole set of cmsis for saml21.
> >
> > I'm sorry if you have the impression that I'm picking on you. That's
> definitely not my intend. Feel invited for a drink the moment we meet in
> person! :)
> >>>
> >>> IMHO we should just select the right cmsis header directly (in
> cpu_conf.h).
> >>> You can't. Not without adding a define like, eh, SAM0_FAM_xyz. If you
> had let me move samd21 to samr21 and add the 'real' samd21, then the three
> samd21, samr21, saml21 would all have had the same structure. And SAM0_FAM
> wouldn't be needed.
> >
> > Well, I think we shouldn't let the vendor headers influence too much on
> how we best organize our platforms.
> >
> > The samr21 is just a variant of samd21, even if Atmel is shipping a whole
> new include tree. A diff shows that they are practically identical.
> >
> > Please take a look at keestux#1.
> >
> > In there I've just copied the four samr21 specific files into the samd21
> cmsis folder, and added the extra case to samd21.h. Works fine as is, saves
> 20k lines of duplicated header files and removes the need for SAM0_FAM.
>
> Keep in mind the maintainability of the files. I don't see any problem with
> vendor headers duplicating their content when it reduces our own
> maintenence effort, so I don't agree with your opinion, Kaspar. But I just
> want this issue to be resolved so that we can focus on more productive
> tasks.
>
> >
> > —
> > You are receiving this because you commented.
> >
> > Reply to this email directly, view it on GitHub, or mute the thread.
>
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/RIOT-OS/RIOT/pull/5743#issuecomment-239609206
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/5743#issuecomment-239611665>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AATYQo4P4UpuPqOpbhdXd2KGW0LlPiI9ks5qfYxegaJpZM4Jhh2G>
> .
>


-- 
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub:
https://github.com/RIOT-OS/RIOT/pull/5743#issuecomment-239612224

PeterKietzmann commented:
@kaspar030 I can give a bit support in detailed testing. Already did a quick test with networking, spi and i2c which worked well.

kaspar030 commented:
Ok, you convinced me that fiddling with the cmsis files or their structure is a bad idea compared to the small overhead of duplicated/identical vendor headers.

Still, @keestux, I would like to get rid of the SAM0_FAM C defines. As I see it, it is only used for selecting the correct cmsis include path (in the makefiles) and include file.

Could you live with creating a ```sam0.h``` which basically looks the same as e.g., ```samr21.h```, but listing all sam0 variants we support, that either includes the correct ```saml21a12b.h``` header or just passes on to ```saml21.h``` and ```samr21.h```?

keestux commented:
@kaspar030 About the sam0.h, yes that seems possible. I'll give it a try.

kYc0o commented:
As in #5525 Murdock doesn't seem to finish here too...

kaspar030 commented:
> As in #5525 Murdock doesn't seem to finish here too...

Murdock is working again, some builds which show "cancelled" will need to be restarted, though.

keestux commented:
Ping

PeterKietzmann commented:
@kaspar030 Murdock was successful. Would take a look at the latest change please?

kaspar030 commented:
I'm getting the endianness redefinition error:

```
/home/kaspar/src/riot.tmp/cpu/sam21_common/include/cmsis/samr21/include/samr21g18a.h:232:0: error: "LITTLE_ENDIAN" redefined [-Werror]
 #define LITTLE_ENDIAN          1

In file included from /usr/arm-none-eabi/include/sys/types.h:67:0,
                 from /usr/arm-none-eabi/include/stdio.h:48,
                 from /home/kaspar/src/riot.tmp/examples/default/main.c:25:
/usr/arm-none-eabi/include/machine/endian.h:17:0: note: this is the location of the previous definition
 #define LITTLE_ENDIAN _LITTLE_ENDIAN

cc1: all warnings being treated as errors
```

keestux commented:
Can you tell me under what circumstances you get that error, because I don't get it (neither does Murdock).

kaspar030 commented:
> Can you tell me under what circumstances you get that error, because I don't get it (neither does Murdock).

I get it on arch linux when building the default example for samr21-xpro. There was a workaround for this in the old samr21 headers.

kaspar030 commented:
> There was a workaround for this in the old samr21 headers.

```
#ifdef LITTLE_ENDIAN
#undef LITTLE_ENDIAN
#endif
```

... in sam0.h helps.

The define is triggered through newlib's "_BSD_VISIBLE",  which seems to be default.

keestux commented:
Yes, I know there was a work around, but I don't want to use workarounds without understanding why.
Plus, why doesn't Murdock have this problem? Why don't I have this problem? Why doesn't Peter have this problem? Weird.

In /usr/arm-none-eabi/include/machine/endian.h there is:

    /*  Code relocated from libc/sys/arm/sys/param.h.  */
    #define BIG_ENDIAN      4321
    #define LITTLE_ENDIAN   1234

and in the ASF include files there is:

    #define LITTLE_ENDIAN          1

That seems a bug in the ASF files, if you ask me. It is up to the compiler to decide if it is little or big endian. And then in endian.h it selects BYTE_ORDER from BIG_ENDIAN or LITTLE_ENDIAN. We shouldn't mess with the definition of LITTLE_ENDIAN.

What do you think, shall I remove the define of LITTLE_ENDIAN all together (by means of a sed command)? It isn't used anywhere in the ASF files.

kaspar030 commented:
> Yes, I know there was a work around, but I don't want to use workarounds
> without understanding why.

Always a good idea.

> Plus, why doesn't Murdock have this problem? Why don't I have this
> problem? Why doesn't Peter have this problem? Weird.

Indeed weird. Makefile.cflags defaults to "-std=gnu99", which enables
"__BSD_VISIBLE" in my newlib version (2.4.0), which in turn sets the
endian define.

What toolchain / newlib version are you on?

I think changing the CFLAGS to "-std=c99" is the way to go here, see #5824.

keestux commented:
I tried -std=gnu99 and -std=c99. Both give no errors with LITTLE_ENDIAN.

I'm on Ubuntu 14.04 with  gcc-arm-none-eabi  4.9.3.2015q3-1trusty1
I don't have newlib installed. (Ubuntu package: libnewlib-arm-none-eabi). Should I?

gebart commented:
I agree that the ASF headers are wrong and it should be defined by the
compiler. I suggest removing the definition in ASF.

gebart commented:
Also, there are differences between tool chains. Murdock is running, as far
as I know, gcc 5.4.0 and newlib from the launchpad gcc-arm-embedded ppa.

keestux commented:
@kaspar030 Please test with the removed LITTLE_ENDIAN.

keestux commented:
Shall I rebase?
Is that common practise for a PR that takes a little longer to get merged?

keestux commented:
Ping

keestux commented:
@kaspar030 Update is available

- rebase, done
- dropped the addition of samd21 (they'll be added with upcoming PRs)
- squashed, not done (I wanted to keep each commit as is, because each commit tells its own story. It will be easier to "blame" if something turned out to introduce bugs, which of course is very unlikely :-) )

kaspar030 commented:
ACK.

immesys commented:
Wow, great work @keestux. I'm also putting together a board based on SAMR21, so I appreciate you cleaning stuff up!

haukepetersen commented:
I know this PR took a long road and I support the effort for a cleaner structure in the Atmel CPU tree of RIOT. But I think here are some/many things that went wrong with this PR! My main issues are:

- why host the samr/d21 header in the sam21_common, while the saml21 header are not? -> needs unification
- most of the commits do more/different things then they say in their commit message
- bad code style was merged (e.g. c++-style comments in `cpu/samd21/periph/gpio.c`
- functionality unrelated to this PR was merged -> adding/changing some `sercom` functionality for periph drivers is completely unrelated to this PR
- functions where not reviewed properly: the newly added `gpio_init_sercom()` does exactly the same than the already existing `gpio_init_mux()` function
- etc...

@kYc0o @PeterKietzmann @kaspar030 I know this PR took a long road, but let's not relax our requirements on code quality and PR structure!

@keestux I really appreciate your effort! Maybe next time try to go in small steps, which makes things so much easier (and hopefully faster) to review and merge.

keestux commented:
@haukepetersen Hmm. You did notice that this already got merged, right?

- About saml21 in sam21_common, but no headers. Preferably saml21 should go in there as well.
- I really try to make a commit do one thing only.
- C++ style comment are not documented as "bad code style". The wiki states: "code shall be C99 compliant", which allows //. This code in particular was copied from Arduino. And in these cases I want to minimize the differences with the original.
- Unrelated functionality was added. Well, without these changes it was not easy to deliver working commits. I did my best to first send in changes with the ultimate goal to have my board supported at some point. It was a battle.
- functions were not reviewed properly? Hmm.
- etc? Keep going. I can handle it :-)

Small steps? Like #4780 ? ;-)

miri64 commented:
> Small steps? Like #4780 ? ;-)

I think what Hauke is trying to say is that though there is no convention as to how big a PR is to be, this is - as a rule of thumb :stuck_out_tongue_winking_eye: - generally not good for review:

![screenshot from 2016-10-07 15-47-56](https://cloud.githubusercontent.com/assets/675644/19192235/e9bb1012-8ca5-11e6-8db9-521c48a192ab.png)

immesys commented:
As an outside, I must agree. Even after I knew that this PR was responsible for #5868, searching through the changes made in the PR took a very long time.

OlegHahm commented:
> C++ style comment are not documented as "bad code style". The wiki states: "code shall be C99 compliant", which allows //. 

Our coding conventions are inherit from the Linux coding conventions. See section 8 in 
https://www.kernel.org/doc/Documentation/CodingStyle.
